// @microsoft/spfx-heft-plugins/lib/plugins/webpackServePlugin/WebpackServePlugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpackServePlugin = void 0;
const path = __importStar(require("path"));
const node_core_library_1 = require("@rushstack/node-core-library");
const express_1 = __importDefault(require("express"));
const ConfigureServe_1 = require("./ConfigureServe");
const InfoPlugin_1 = require("./webpackPlugins/InfoPlugin");
const MetricsWebpackPlugin_1 = require("./webpackPlugins/MetricsWebpackPlugin");
const CumulativeManifestProcessor_1 = require("../../spfxManifests/cumulativeManifestProcessor/CumulativeManifestProcessor");
const constants_1 = require("../../utilities/constants");
/**
 * @public
 */
class WebpackServePlugin {
    constructor() {
        this.pluginName = "spfx-webpack-serve-plugin" /* SpfxWebpackServePlugin */;
    }
    /**
     * Generate the Webpack Dev Server configuration for SPFx component development
     */
    apply(heftSession, heftConfiguration) {
        const logger = heftSession.requestScopedLogger('spfx-webpack-serve-plugin');
        const terminal = logger.terminal;
        const { buildFolder } = heftConfiguration;
        const disableDebugLoaderParameter = heftSession.commandLine.registerFlagParameter({
            associatedActionNames: ['start'],
            parameterLongName: '--no-loader',
            description: 'If specified, the debug loader query param will be omitted from the debug url.'
        });
        const debugLoaderPathParameter = heftSession.commandLine.registerStringParameter({
            associatedActionNames: ['start'],
            parameterLongName: '--loader-path',
            description: 'Optionally specify the url for a debug loader query param.',
            argumentName: 'LOADER_URL'
        });
        heftSession.hooks.build.tap("spfx-webpack-serve-plugin" /* SpfxWebpackServePlugin */, (build) => {
            if (build.properties.serveMode) {
                build.hooks.bundle.tap("spfx-webpack-serve-plugin" /* SpfxWebpackServePlugin */, (bundle) => {
                    bundle.hooks.configureWebpack.tapPromise("spfx-webpack-serve-plugin" /* SpfxWebpackServePlugin */, async (untypedConfig) => {
                        let config = untypedConfig;
                        if (!config) {
                            throw new Error('Serve mode is expecting a webpack config to use webpack dev server.');
                        }
                        // Webpack Dev Server will only consider the first config if multiple are defined.
                        if (Array.isArray(config)) {
                            config = config[0];
                        }
                        const serve = new ConfigureServe_1.ConfigureServe(heftConfiguration, terminal);
                        const serveConfig = serve.generateServeConfig();
                        // Project serve paths
                        const projectServeMap = new Map();
                        const { contentBasePublicPath = '/' } = serveConfig;
                        if (serveConfig.contentBase) {
                            if (Array.isArray(serveConfig.contentBase)) {
                                for (const contentPath of serveConfig.contentBase) {
                                    projectServeMap.set(path.resolve(buildFolder, contentPath), contentBasePublicPath);
                                }
                            }
                            else {
                                projectServeMap.set(path.resolve(buildFolder, serveConfig.contentBase), contentBasePublicPath);
                            }
                        }
                        else {
                            projectServeMap.set(path.resolve(buildFolder), contentBasePublicPath);
                        }
                        // Additional serve paths
                        const tabTestServeMap = await this._getTabTestServeMapAsync(buildFolder);
                        const dependencyServeMap = this._getDependencyServeMap(buildFolder, terminal);
                        const serveMap = new Map([
                            ...projectServeMap,
                            ...tabTestServeMap,
                            ...dependencyServeMap
                        ]);
                        const metricsCollector = heftSession.metricsCollector;
                        // Webpack Dev Server config
                        config.devServer = {
                            contentBase: false,
                            publicPath: '/dist',
                            port: serveConfig.port,
                            hot: true,
                            historyApiFallback: false,
                            compress: true,
                            disableHostCheck: true,
                            headers: {
                                'Access-Control-Allow-Origin': '*',
                                'Access-Control-Allow-Methods': 'HEAD, GET, OPTIONS'
                            },
                            stats: {
                                assets: false,
                                chunks: false,
                                modules: false,
                                warningsFilter: [/export .* was not found in/]
                            },
                            before: (app, server, compiler) => {
                                if (serveConfig.apiMap) {
                                    for (const [route, handler] of Object.entries(serveConfig.apiMap)) {
                                        // Don't include universal route handlers.
                                        if (route.startsWith('*')) {
                                            continue;
                                        }
                                        app.get(route, handler);
                                    }
                                }
                                app.use(serve.logRequestsMiddleware());
                            },
                            after: (app, server, compiler) => {
                                for (const [contentPath, route] of serveMap) {
                                    app.use(route, express_1.default.static(contentPath, { index: false }));
                                }
                            },
                            open: !!serveConfig.initialPage,
                            openPage: serveConfig.initialPage,
                            writeToDisk: true
                        };
                        // Webpack Dev Server plugins
                        config.plugins = [
                            ...(config.plugins ? config.plugins : []),
                            new InfoPlugin_1.InfoPlugin({
                                terminal: terminal,
                                port: serveConfig.port,
                                tempFolderName: constants_1.DEFAULT_TEMP_FOLDER,
                                setDebugLoader: !disableDebugLoaderParameter.value,
                                debugLoaderPath: debugLoaderPathParameter.value
                            }),
                            new MetricsWebpackPlugin_1.MetricsWebpackPlugin(metricsCollector)
                        ];
                        return config;
                    });
                });
            }
        });
    }
    async _getTabTestServeMapAsync(buildFolder) {
        const tabTestServeMap = new Map();
        const tabSrcPath = `${buildFolder}/src/tab/`;
        if (await node_core_library_1.FileSystem.existsAsync(tabSrcPath)) {
            tabTestServeMap.set(path.resolve(buildFolder, 'temp'), '/');
            tabTestServeMap.set(path.resolve(buildFolder, 'lib', 'tab'), '/tab/');
            // TAB tests are expected to be located at that /tab/ root. Therefore, we must create the effect of
            // a flattened tab directory by serving content from tab subdirectories on the /tab/ root.
            const tabFolderContents = await node_core_library_1.FileSystem.readFolderItemsAsync(tabSrcPath);
            for (const tabFolderContent of tabFolderContents) {
                if (tabFolderContent.isDirectory()) {
                    tabTestServeMap.set(path.resolve(buildFolder, 'lib', 'tab', tabFolderContent.name), '/tab/');
                }
            }
        }
        return tabTestServeMap;
    }
    _getDependencyServeMap(buildFolder, terminal) {
        const dependencyServeMap = new Map();
        const referencedProjects = this._getReferencedProjects(buildFolder, terminal);
        for (const versionMap of Object.values(referencedProjects)) {
            for (const referencedProject of Object.values(versionMap)) {
                if (referencedProject.packageName && !referencedProject.isAssembly) {
                    const projectOutputPath = path.dirname(referencedProject.manifestPath);
                    if (referencedProject.manifestPath) {
                        dependencyServeMap.set(projectOutputPath, '/');
                    }
                    else {
                        terminal.writeWarning(referencedProject.packageName + ' not found.');
                    }
                }
            }
        }
        return dependencyServeMap;
    }
    _getReferencedProjects(buildFolder, terminal) {
        const cumulativeManifestProcessor = new CumulativeManifestProcessor_1.CumulativeManifestProcessor({
            explicitInclude: [constants_1.SP_LOADER_ASSEMBLY_ID],
            rootPath: buildFolder,
            tempFolderName: constants_1.DEFAULT_TEMP_FOLDER,
            distFolderName: constants_1.DEFAULT_OUTPUT_FOLDER,
            includeAssemblies: true,
            terminal: terminal
        });
        const referencedProjects = cumulativeManifestProcessor.discoverManifests(buildFolder, CumulativeManifestProcessor_1.DependencyDiscoveryMode.deepSparseIgnoreFirstProject);
        return referencedProjects;
    }
}
exports.WebpackServePlugin = WebpackServePlugin;
exports.default = new WebpackServePlugin(); // tslint:disable-line:export-name
