// @rushstack/rush-amazon-s3-build-cache-plugin/lib/RushAmazonS3BuildCachePlugin.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RushAmazonS3BuildCachePlugin = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
const AmazonS3BuildCacheProviderModule = node_core_library_1.Import.lazy('./AmazonS3BuildCacheProvider', require);
const PLUGIN_NAME = 'AmazonS3BuildCachePlugin';
/**
 * @public
 */
class RushAmazonS3BuildCachePlugin {
    constructor() {
        this.pluginName = PLUGIN_NAME;
    }
    apply(rushSession, rushConfig) {
        rushSession.hooks.initialize.tap(PLUGIN_NAME, () => {
            rushSession.registerCloudBuildCacheProviderFactory('amazon-s3', (buildCacheConfig) => {
                const { amazonS3Configuration } = buildCacheConfig;
                return new AmazonS3BuildCacheProviderModule.AmazonS3BuildCacheProvider({
                    s3Region: amazonS3Configuration.s3Region,
                    s3Bucket: amazonS3Configuration.s3Bucket,
                    s3Prefix: amazonS3Configuration.s3Prefix,
                    isCacheWriteAllowed: !!amazonS3Configuration.isCacheWriteAllowed
                }, rushSession);
            });
        });
    }
}
exports.RushAmazonS3BuildCachePlugin = RushAmazonS3BuildCachePlugin;
