// @rushstack/module-minifier-plugin/lib/LocalMinifier.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalMinifier = void 0;
const MinifySingleFile_1 = require("./terser/MinifySingleFile");
require("./OverrideWebpackIdentifierAllocation");
/**
 * Minifier implementation that minifies code on the main thread.
 * @public
 */
class LocalMinifier {
    constructor(options) {
        const { terserOptions = {} } = options || {};
        this._terserOptions = {
            ...terserOptions,
            output: terserOptions.output
                ? {
                    ...terserOptions.output
                }
                : {}
        };
        this._resultCache = new Map();
    }
    /**
     * Transform that invokes Terser on the main thread
     * @param request - The request to process
     * @param callback - The callback to invoke
     */
    minify(request, callback) {
        const { hash } = request;
        const cached = this._resultCache.get(hash);
        if (cached) {
            return callback(cached);
        }
        (0, MinifySingleFile_1.minifySingleFile)(request, this._terserOptions)
            .then((result) => {
            this._resultCache.set(hash, result);
            callback(result);
        })
            .catch((error) => {
            // This branch is here to satisfy the no-floating-promises lint rule
            callback({
                error: error,
                code: undefined,
                map: undefined,
                hash
            });
        });
    }
}
exports.LocalMinifier = LocalMinifier;
