// @microsoft/sp-webpart-base/lib-commonjs/common/Flights.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOptionalSupportedHostsFilteringEnabled = exports.is1PIsolationEnabled = exports.isCallGetClientSideComponentsByComponentTypeEnabled = exports.isIsolatedAceEnabled = exports.isVivaDashboardImagePerfEnabled = exports.isMeeDashboardEnabled = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
// tslint:disable:export-name
function isMeeDashboardEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(638 /* MEEDashboard */);
}
exports.isMeeDashboardEnabled = isMeeDashboardEnabled;
function isVivaDashboardImagePerfEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60338 /* VivaDashboardImagePerf */);
}
exports.isVivaDashboardImagePerfEnabled = isVivaDashboardImagePerfEnabled;
function isIsolatedAceEnabled() {
    var isSafari = navigator.userAgent.indexOf('Chrome') === -1 && navigator.userAgent.indexOf('Safari') > -1;
    return (sp_core_library_1._SPFlight.isEnabled(60290 /* SPPPlatIsolatedAce */) &&
        (!isSafari || sp_core_library_1._SPFlight.isEnabled(60524 /* SPPPlatSafariBridgeIsolation */)));
}
exports.isIsolatedAceEnabled = isIsolatedAceEnabled;
function isCallGetClientSideComponentsByComponentTypeEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60309 /* GetClientSideComponentsByComponentType */);
}
exports.isCallGetClientSideComponentsByComponentTypeEnabled = isCallGetClientSideComponentsByComponentTypeEnabled;
function is1PIsolationEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60400 /* SPPPlatFirstPartyIsolation */);
}
exports.is1PIsolationEnabled = is1PIsolationEnabled;
function isOptionalSupportedHostsFilteringEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60079 /* OptionalSupportedHostsFiltering */);
}
exports.isOptionalSupportedHostsFilteringEnabled = isOptionalSupportedHostsFilteringEnabled;
