// @microsoft/sp-webpart-base/lib-commonjs/common/KillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isClearLoadingIndicatorKSActivated = exports.isOpenPropPaneInSeparateIframeKSActivated = exports.isSupportThemesInIsolatedWebPartKSActivated = exports.isAutomaticThemeSupportInBaseClientSideWebPartKSActivated = exports.isIsolatedWebPartTabStopKSActivated = exports.isWebPartZeroWidthNotStoredInCacheKSActivated = exports.growDashboardWebPartACEKSActivated = exports.useTeamsSDKFromHostedAppKSActivated = exports.fixACERequiringWPinSupportedHostsKSActivated = exports.fixIsolatedDomainsOnSubWebsKSActivated = exports.isAllowEmptyPropertyNameKSActivated = exports.isLogWebPartLoadFailureKSActivated = exports.isAddTeamsHostDataKsActivated = exports.clearTimerWhenRenderFailKSActivated = exports.isAllowProactiveRenderFailureKsActivated = exports.isClearIntervalTasksKsActivated = exports.isIsolatedWebPartParentHostLifeCycleKsActivated = exports.isMoveAsyncTimerKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
var resolveRenderTimeGuid = sp_core_library_1.Guid.parse('d2b2d5e6-4bd3-40f7-b49f-782f95ae1ae1');
var KillSwitches = /** @class */ (function () {
    function KillSwitches() {
    }
    KillSwitches.isResolveRenderTimeActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(resolveRenderTimeGuid, '2020/04/15', 'Render complete time logged synchronously');
    };
    KillSwitches.isAddSyncRenderTimeKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('57437895-9f0b-42f0-9c9e-aa5cea5286dc'), '05/09/2020', 'Add async Render Time');
    };
    KillSwitches.isExclusiveRenderTimeForBaseCollectionWebPartKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('d84e2f1c-760e-4fb3-9589-95de35f2ddc1'), '06/09/2020', 'Add exclusive render time for base collection web part');
    };
    KillSwitches.isRejectLegacyFabricErrorKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('10f94589-c5da-4d9c-8b4e-040a54822b18'), '06/05/2020', 'Reject load legacy fabric css promise if loading legacy css fails');
    };
    KillSwitches._isWebpartExperimentKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('4cfbf54a-8bda-43ee-8599-e62f0d1ae9ee'), '2020/06/11', 'Log web part perf experiments and flights');
    };
    KillSwitches.isTeamsContextKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('a5daffb0-652e-4ee1-a638-e290089e1de7'), '09/11/2020', 'Refactor a method of IWebPartHost as async, as part of PR to enable teams context.');
    };
    return KillSwitches;
}());
exports.default = KillSwitches;
function isMoveAsyncTimerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('28054d9c-1d49-4ff2-a20d-9c8734a53ca4' /* '08/03/2020', 'Starting async timer before render' */);
}
exports.isMoveAsyncTimerKSActivated = isMoveAsyncTimerKSActivated;
function isIsolatedWebPartParentHostLifeCycleKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ad020f8e-7591-407a-b812-2f8f111b9014'
    /** 08/19/2020, 'updates parent host life cycle when isolated web part life cycle updates */
    );
}
exports.isIsolatedWebPartParentHostLifeCycleKsActivated = isIsolatedWebPartParentHostLifeCycleKsActivated;
function isClearIntervalTasksKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1c24a303-960c-46ad-8602-385365076bf4'
    /** 11/16/2020, 'Clear interval task when disposed' */
    );
}
exports.isClearIntervalTasksKsActivated = isClearIntervalTasksKsActivated;
function isAllowProactiveRenderFailureKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e50c4ad7-27e6-4665-abde-f8687023ce6b'
    /* '1/11/2021', 'Allow proactive render failure for async render web part.' */
    );
}
exports.isAllowProactiveRenderFailureKsActivated = isAllowProactiveRenderFailureKsActivated;
function clearTimerWhenRenderFailKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('514fb476-3130-4a90-98bd-493451902251' /* '07/06/2021', 'Clear timer when async render failed' */);
}
exports.clearTimerWhenRenderFailKSActivated = clearTimerWhenRenderFailKSActivated;
// This is also in extensions when it comes time to clean up'
function isAddTeamsHostDataKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('2a3f9090-e367-4a11-908d-592a69f875a8'
    /* '3/11/2021', 'log extra telemetry when components are hosted in teams. */
    );
}
exports.isAddTeamsHostDataKsActivated = isAddTeamsHostDataKsActivated;
function isLogWebPartLoadFailureKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d73e81a7-648f-45e4-b7c5-09dbcba6ddc7' /* '03/30/2021', 'Log web part alias when web part load failed.' */);
}
exports.isLogWebPartLoadFailureKSActivated = isLogWebPartLoadFailureKSActivated;
function isAllowEmptyPropertyNameKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d39e74cc-8615-40e7-9f1b-bd6ad16dda59'
    /* '07/08/2021', 'Allow but do not persist properties with empty property name' */
    );
}
exports.isAllowEmptyPropertyNameKSActivated = isAllowEmptyPropertyNameKSActivated;
function fixIsolatedDomainsOnSubWebsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ee03b10a-9b6e-40bc-bd08-83088a0ee756'
    /* '07/23/2021', 'Use the web path instead of site path so isolated webparts load on sub webs' */
    );
}
exports.fixIsolatedDomainsOnSubWebsKSActivated = fixIsolatedDomainsOnSubWebsKSActivated;
function fixACERequiringWPinSupportedHostsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6c94e4d7-b951-48c7-b282-318ad7525e96'
    /* '10/5/2021', 'Temporary fix to stop requiring ACEs have SharePointWebPart in their supportedHosts array' */
    );
}
exports.fixACERequiringWPinSupportedHostsKSActivated = fixACERequiringWPinSupportedHostsKSActivated;
function useTeamsSDKFromHostedAppKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c7acb771-f16d-4cab-989d-ab6200215d94'
    /* '8/9/2021', 'use teams host' */
    );
}
exports.useTeamsSDKFromHostedAppKSActivated = useTeamsSDKFromHostedAppKSActivated;
function growDashboardWebPartACEKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e215de04-57bf-4cc3-af49-4ebf4e2a5780'
    /* '9/9/2021', 'Dashboard web parts with 6 cards need to grow to fit the space' */
    );
}
exports.growDashboardWebPartACEKSActivated = growDashboardWebPartACEKSActivated;
function isWebPartZeroWidthNotStoredInCacheKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e215de04-57bf-4cc3-af49-4ebf4e2a5780'
    /* '9/22/2021', 'Web part width cache manager will not store 0 width value.' */
    );
}
exports.isWebPartZeroWidthNotStoredInCacheKSActivated = isWebPartZeroWidthNotStoredInCacheKSActivated;
function isIsolatedWebPartTabStopKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d8618918-ac55-4979-b9d4-be944a4feb95'
    /* '10/25/2021', 'adds tabindex=0 to the isolated webpart iframe' */
    );
}
exports.isIsolatedWebPartTabStopKSActivated = isIsolatedWebPartTabStopKSActivated;
function isAutomaticThemeSupportInBaseClientSideWebPartKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('42f2d6d0-5ea1-4df9-8f02-8b68ff4a7d4d'
    /* '9/24/2021', 'automatic theme support' */
    );
}
exports.isAutomaticThemeSupportInBaseClientSideWebPartKSActivated = isAutomaticThemeSupportInBaseClientSideWebPartKSActivated;
function isSupportThemesInIsolatedWebPartKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5475f644-fe11-4f4c-b9c7-e5850a69941a'
    /* '10/22/2021', 'Support for themes in isolated web parts' */
    );
}
exports.isSupportThemesInIsolatedWebPartKSActivated = isSupportThemesInIsolatedWebPartKSActivated;
function isOpenPropPaneInSeparateIframeKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c9b8e7c2-b8a3-4b3c-a8d7-e9b9c9a9f9c4'
    /* '11/01/2021', 'Open property pane in separate iframe' */
    );
}
exports.isOpenPropPaneInSeparateIframeKSActivated = isOpenPropPaneInSeparateIframeKSActivated;
function isClearLoadingIndicatorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d21f0e82-8554-4a61-ae3e-03015288aefc'
    /* '01/04/2022', 'clearLoadingIndicator callback' */
    );
}
exports.isClearLoadingIndicatorKSActivated = isClearLoadingIndicatorKSActivated;
