// @microsoft/sp-loader/lib-commonjs/tiny/common/createRootScope.js
"use strict";
// tslint:disable:export-name
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRootScope = void 0;
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_diagnostics_1 = require("@microsoft/sp-diagnostics");
var sp_dynamic_data_1 = require("@microsoft/sp-dynamic-data");
var sp_http_1 = require("@microsoft/sp-http");
var sp_page_context_1 = require("@microsoft/sp-page-context");
var flights_1 = require("./flights");
var killSwitches_1 = require("./killSwitches");
/**
 * Create a root service scope and initialize with SPFx services:
 *   - PageContext
 *   - DynamicDataManager
 *   - HttpClient
 *   - SPHttpClient
 *   - GraphHttpClientContext
 *   - DigestCache
 *
 * @param pageContext - the current SPPageContext
 *
 * @returns - the root scope
 */
function createRootScope(pageContext) {
    var root = sp_core_library_1.ServiceScope.startNewRoot();
    root.provide(sp_diagnostics_1._logSourceServiceKey, sp_diagnostics_1._LogSource.create('RootServiceScope'));
    root.createDefaultAndProvide(sp_page_context_1.PageContext.serviceKey);
    root.createDefaultAndProvide(sp_dynamic_data_1._DynamicDataUtilities.IDynamicDataManagerServiceKey);
    root.createDefaultAndProvide(sp_http_1.HttpClient.serviceKey);
    var spHttpClient = root.createDefaultAndProvide(sp_http_1.SPHttpClient.serviceKey);
    var graphContext = root.createDefaultAndProvide(sp_http_1._GraphHttpClientContext.serviceKey);
    var digestCache = root.createDefaultAndProvide(sp_http_1.DigestCache.serviceKey);
    root.finish();
    initializeGraphHttpClient(pageContext, spHttpClient, graphContext);
    initializeDigestCache(pageContext, digestCache);
    return root;
}
exports.createRootScope = createRootScope;
function initializeGraphHttpClient(pageContext, spHttpClient, graphContext) {
    var aadInstanceUrl = pageContext.aadInstanceUrl, aadSessionId = pageContext.aadSessionId, aadTenantId = pageContext.aadTenantId, aadUserId = pageContext.aadUserId, CorrelationId = pageContext.CorrelationId, env = pageContext.env, isAnonymousGuestUser = pageContext.isAnonymousGuestUser, isExternalGuestUser = pageContext.isExternalGuestUser, msGraphEndpointUrl = pageContext.msGraphEndpointUrl, spfx3rdPartyServicePrincipalId = pageContext.spfx3rdPartyServicePrincipalId, spfxOBOFlowEnabled = pageContext.spfxOBOFlowEnabled, userPrincipalName = pageContext.userPrincipalName, webAbsoluteUrl = pageContext.webAbsoluteUrl, webServerRelativeUrl = pageContext.webServerRelativeUrl;
    if (sp_core_library_1.Environment.type !== sp_core_library_1.EnvironmentType.Local) {
        graphContext.initialize(webServerRelativeUrl, msGraphEndpointUrl, webAbsoluteUrl);
    }
    var defaultAadConfig = {
        aadInstanceUrl: aadInstanceUrl,
        aadSessionId: aadSessionId,
        aadTenantId: aadTenantId,
        aadUserId: aadUserId,
        redirectUri: window.location.origin + "/_forms/" + sp_http_1._AadConstants.SPFX_SINGLE_SIGN_ON_REPLY_URL,
        servicePrincipalId: spfx3rdPartyServicePrincipalId,
        spRequestGuid: CorrelationId,
        userPrincipalName: isAnonymousGuestUser || isExternalGuestUser ? undefined : userPrincipalName
    };
    var oboConfig;
    defaultAadConfig.aadInstanceUrl =
        !env || env.toLowerCase() !== 'edog'
            ? 'https://login.microsoftonline.com'
            : 'https://login.windows-ppe.net';
    if (((spfxOBOFlowEnabled || !killSwitches_1.isRemoveSPFxOBOFlowEnabledFlagKSActivated()) &&
        sp_core_library_1._BrowserUtilities.isMobileWebView()) ||
        sp_core_library_1._BrowserUtilities.isWebViewHosted() ||
        (flights_1.isSafariAuthPatchEnbled() && /.*AppleWebKit.*Safari/.test(navigator.userAgent))) {
        oboConfig = { spHttpClient: spHttpClient, serverRelativeUrl: webAbsoluteUrl };
    }
    try {
        sp_http_1._AadTokenProviders._initialize(new sp_http_1.AadTokenProvider(defaultAadConfig, oboConfig), tslib_1.__assign(tslib_1.__assign({}, defaultAadConfig), { servicePrincipalId: sp_http_1._AadConstants.PRE_AUTHORIZED_APP_PRINCIPAL_ID }));
    }
    catch (e) {
        sp_diagnostics_1._TraceLogger.logVerbose(sp_diagnostics_1._LogSource.create('RootServiceScope'), 'AadTokenProviders: Failed to initialize');
    }
}
function initializeDigestCache(pageContext, digestCache) {
    var formDigestTimeoutSeconds = pageContext.formDigestTimeoutSeconds, formDigestValue = pageContext.formDigestValue, serverTime = pageContext.serverTime, webAbsoluteUrl = pageContext.webAbsoluteUrl, webServerRelativeUrl = pageContext.webServerRelativeUrl;
    // Value of serverTime is same as what comes as part of formDigestValue
    // but is in locale neutral ISO 8601 format. So it will get correctly
    // parsed by Date class irrespective of client locale.
    // serverTime is accurate to the order of ms, while DateTime which comes
    // as part of formDigestValue is trimmed to order of seconds. Subtract
    // 30s from expirationTimeStamp to avoid any timing errors b/w server and client
    var expirationTimestamp = new Date(serverTime).getTime() + 1000 * formDigestTimeoutSeconds - 30000;
    for (var _i = 0, _a = [webAbsoluteUrl, webServerRelativeUrl]; _i < _a.length; _i++) {
        var url = _a[_i];
        digestCache.addDigestToCache(url, formDigestValue, expirationTimestamp);
    }
}
