// @microsoft/sp-loader/lib-commonjs/tiny/common/killSwitches.js
"use strict";
// tslint:disable:export-name
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRemoveSPFxOBOFlowEnabledFlagKSActivated = exports.isLogSuccessOfTryGetManifestsKillSwitchActivated = exports.isLoadComponentDependenciesKillSwitchActivated = exports.isRejectOnNotFoundManifestsKillSwitchActivated = exports.isRequestManifestKillSwitchActivated = exports.isDisable3PCodeKillswitchActivated = exports.isSkipFirstPartyTinyLogKillswtichActivated = exports.isSkipFirstPartyLoaderLogKillswtichActivated = exports.isDangerouslyEnableDebugKSActivated = exports.isReactProfilerKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isReactProfilerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('86a7f111-fd30-473c-ae8e-2aebd51e76fb'
    /* '04/13/2020', 'Add the toggle that can enable the react profiler on the existence of specified query' */
    );
}
exports.isReactProfilerKSActivated = isReactProfilerKSActivated;
function isDangerouslyEnableDebugKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e0187f01-1210-45a0-9865-716fedd0630a'
    /* '07/17/19', 'Enable debug when using debug loader' */
    );
}
exports.isDangerouslyEnableDebugKSActivated = isDangerouslyEnableDebugKSActivated;
function isSkipFirstPartyLoaderLogKillswtichActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8f96cdcd-e215-4a2b-870e-9857af6b3706'
    /* '10/01/20', 'Stop logging loadComponent QoS in SPComponentLoader' */
    );
}
exports.isSkipFirstPartyLoaderLogKillswtichActivated = isSkipFirstPartyLoaderLogKillswtichActivated;
function isSkipFirstPartyTinyLogKillswtichActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a67ffebb-e928-40d1-b6a9-4d544b9f0618'
    /* '10/14/20', 'Stop logging loadComponent QoS in RequireLoader' */
    );
}
exports.isSkipFirstPartyTinyLogKillswtichActivated = isSkipFirstPartyTinyLogKillswtichActivated;
function isDisable3PCodeKillswitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d326e3b8-c718-43c5-b920-a2b2ee0ea0fb'
    /* '11/13/20', 'Allow a query string to stop loading 3rd party code' */
    );
}
exports.isDisable3PCodeKillswitchActivated = isDisable3PCodeKillswitchActivated;
function isRequestManifestKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1ae2c8f3-eaab-4d49-9a8e-0a4f9e4fe4d5'
    /* '12/18/20', 'Monitor how often we have to fetch manifests' */
    );
}
exports.isRequestManifestKillSwitchActivated = isRequestManifestKillSwitchActivated;
function isRejectOnNotFoundManifestsKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('40391ecf-2963-4b4d-b6cf-a08365509a80'
    /* '12/18/20', 'Reject the manifest fetch if we don't return an actual manifest' */
    );
}
exports.isRejectOnNotFoundManifestsKillSwitchActivated = isRejectOnNotFoundManifestsKillSwitchActivated;
function isLoadComponentDependenciesKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('51f950cd-c0bf-4927-a126-bb8283026ea9'
    /* '12/18/20', 'Add logging to determine which dependency failed to load for a dependent component' */
    );
}
exports.isLoadComponentDependenciesKillSwitchActivated = isLoadComponentDependenciesKillSwitchActivated;
function isLogSuccessOfTryGetManifestsKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('87c8b069-0111-40b1-9700-b9905283d324'
    /* '12/21/20', 'Add logging to so we get request id's on succcess for try get manifests' */
    );
}
exports.isLogSuccessOfTryGetManifestsKillSwitchActivated = isLogSuccessOfTryGetManifestsKillSwitchActivated;
function isRemoveSPFxOBOFlowEnabledFlagKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c3ea6e02-65a6-4991-8528-dce56cb27b00'
    /* '8/23/21', 'Remove checking spfxOBOFlowEnabled flag */
    );
}
exports.isRemoveSPFxOBOFlowEnabledFlagKSActivated = isRemoveSPFxOBOFlowEnabledFlagKSActivated;
